commit 8ddad4c848747bd7134d37700cfcb00641e983b5
Author: Anthony Xu <anthony.xu@intel.com>
Date:   Tue Jun 7 16:23:23 2016 -0700

    execute apic reset in vcpu qemu thread to avoid some race conditions

diff --git a/hw/i386/pc.c b/hw/i386/pc.c
index d270c27..6ddd89c 100644
--- a/hw/i386/pc.c
+++ b/hw/i386/pc.c
@@ -1493,26 +1493,31 @@ static void pc_machine_initfn(Object *obj)
                              pc_machine_set_nvdimm, &error_abort);
 }
 
+static void do_cpu_apic_post_reset(void *arg)
+{
+    CPUState *cs = arg;
+    X86CPU *cpu = X86_CPU(cs);
+    /* Reset APIC after devices have been reset to cancel
+     * any changes that qemu_devices_reset() might have done.
+     */
+    if (cpu->apic_state) {
+        device_reset(cpu->apic_state);
+    }
+    if (boot_info.protected_mode && cpu_is_bsp(cpu)) {
+        reset_cpu(&cpu->env);
+    }
+}					
+
+
 static void pc_machine_reset(void)
 {
     CPUState *cs;
-    X86CPU *cpu;
 
     qemu_devices_reset();
 
-    CPU_FOREACH(cs) {
-        cpu = X86_CPU(cs);
 
-        /* Reset APIC after devices have been reset to cancel
-         * any changes that qemu_devices_reset() might have done.
-         */
-        if (cpu->apic_state) {
-            device_reset(cpu->apic_state);
-        }
-
-        if (boot_info.protected_mode && cpu_is_bsp(cpu)) {
-            reset_cpu(&cpu->env);
-        }
+    CPU_FOREACH(cs) {
+        async_run_on_cpu(cs, do_cpu_apic_post_reset, cs);
     }
 
     if (boot_info.protected_mode) {
